src/json.hpp

```diff
@@ -2896,11 +3120,35 @@ class basic_json
 
     @since version 1.0.0
     */
-    template<typename ValueType, typename std::enable_if<
-                 not std::is_pointer<ValueType>::value, int>::type = 0>
-    ValueType get() const
-    {
-        return get_impl(static_cast<ValueType*>(nullptr));
+    template <typename ValueType,
+              enable_if_t<
+                  not std::is_pointer<ValueType>::value,
+                  int> = 0>
+    auto get() const
+        -> decltype(this->get_impl(static_cast<ValueType *>(nullptr))) {
+      return get_impl(static_cast<ValueType *>(nullptr));
+    }
+
+    template <typename T,
+              enable_if_t<detail::has_from_json<
+                  JSONSerializer, basic_json, uncvref_t<T>>::value 
+                              and not (detail::is_compatible_object_type<object_t, uncvref_t<T>>::value or
+                                                                                        detail::is_compatible_array_type<basic_json_t, uncvref_t<T>>::value or
+                                                                                        detail::is_compatible_float_type<number_float_t, uncvref_t<T>>::value or
+                                                                                        detail::is_compatible_integer_type<number_integer_t, uncvref_t<T>>::value or
+                                                                                        detail::is_compatible_unsigned_integer_type<number_unsigned_t, uncvref_t<T>>::value or
```

Comment:
Same here, define a `detail::is_compatible_basic_json_type` (or similar) that abstracts all this and makes it more readable.